docs/design/README.md

```diff
@@ -0,0 +1,46 @@
+# Design of Node Controller Manager
+
+The design of the Node Controller Manager is influenced by the Kube Controller Manager, where-in mutiple sub-controllers are used to manage the kubernetes clients.
+
+## Design Principles
+
+It's designed to run in master plane of a kubernetes cluster. It follows the best principles and practices of writing controllers, including, but not limited to:
+
+- Reusing code from kube-controller-manager
+- leader election to allow HA deployments of the controller
+- `workqueues` and multiple thread-workers
+- `SharedInformers` that limit to minimum network calls, deserialization and provide helpful create/update/delete events for resources
+- rate-limiting to allow back-off in case of network outages and general instability of other cluster components
+- sending events to respected resources for easy debugging and overview
+- Prometheus metrics, health and (optional) profiling endpoints
+
+## Objects of Node Controller Manager
+
+Node Controller Manager makes use of 4 CRD objects and 1 Kubernetes secret object to manage machines. They are as follows,
+1. Instance-class: A template that contains cloud provider specific details used to create instances.
+1. Instance: Node Controller Manager's object that actually represents an VM.
+1. Instance-set: A group of instances managed by the Node Controller Manager. 
+1. Instance-deploy: A group of instance-sets managed by the Node Controller Manager to allow updation of instances.
+1. Secret: A kubernetes secret that stores cloudconfig (initialization scripts used to create VMs) and cloud specific credentials
+
+## Components of Node Controller Manager
+
+Node Controller Manager is made up of 3 sub-controllers as of now. They are -
+1. Instance Controller (IC): Used to create/update/delete instances. It is the only controller which actually talks to the cloud providers.
+1. Instance Set Controller (ISC): Used to manage instance-sets. This controller makes sure that desired number of instances are always up and running healthy.
+1. Instance Deploy Controller (IDC): Used for updation of instances from one version to another by manipulating the instance-set objects. 
```

Comment:
s/for updation of/to update/

Reply:
Done.
https://github.com/gardener/node-controller-manager/commit/95c68a4c3f483a5f770c7de61f12b25e5411f3c0#diff-8f28081abb21e65fba5a0b0b29404c78R31